rfc/store-deliver.md

```diff
@@ -0,0 +1,310 @@
+# extend store/* protocol with store/deliver 
+
+## Authors
+
+- [Vasco Santos], [Protocol Labs]
+
+## Abstract
+
+`store/*` protocol is missing verifiability for content actually uploaded to the provided presigned URL. This factor together with intention to create a proof of delivery makes critical the introduction of a new capability in the `store/*` namespace. In this RFC, we propose `store/deliver` and `store/confirm` capabilities.
+
+## Introduction
+
+When an agent wants to store some content, agent invokes `store/add` to declare this intention. The `store/add` invocation MUST include the CID of the content to be stored. An invocation example:
+
+```json
+{
+  "can": "store/add",
+  "with": "did:key:abc...",
+  "nb": {
+    "link": "bag...",
+    "size": 1234
+  }
+}
+```
+
+If the service is receptive to store this data, a presigned URL is returned in the receipt of `store/add` invocation, so that the client can store this content. The presigned URL includes a hash derived from the content CID to guarantee the written bytes match that CID. A receipt example:
+
+```json
+{
+  "ran": "bafy...storeAdd",
+  "out": {
+    "ok": {
+      "status": "upload",
+      "url": "https://...",
+      "link": "bag...",
+      "allocated": 1234
+    }
+  },
+  "fx": {
+    "join": null,
+    "fork": []
+  },
+}
+```
+
+Note that content requested to store, may already be stored. If that is the case, the service may issue receipt with `"status": "done"`.
+
+On their schedule, the agent can post the bytes using the presigned URL. However, both the agent and the service never sign a receipt stating that this content is now stored. This is where `store/deliver` and `store/confirm` can help. In short, we can see `store/deliver` as a way for the agent to sign that they did the work, and `store/confirm` for the service to sign that the bytes were received, stored and can now be retrieved.
+
+While not relevant for verifiability, it is also important pointing out that the service MAY need to run some data computation, issue billing events or write indexes for retrievability asynchronously. Today, web3.storage operated store service relies on "S3 Bucket events" to achieve this, but it should not be assumed that all stores web3.storage will rely on will provide such features.
+
+## New capabilities
+
+### `store/deliver` capability
+
+After the agent handled the asynchronous task of posting the bytes, the agent can invoke `store/deliver` to notify the service that the bytes were delivered. This invocation in the future MAY include a challenge for the service based on the posted data.
```

Comment:
Point I'm trying to make is that invocation spec has [promise pipelines](https://github.com/ucan-wg/invocation?tab=readme-ov-file#await) as a means for forming dependencies between tasks. Ideally we would embrace those so task coordination is captured as opposed to having them described in just a spec.

Reply:
@Gozala I definitely see this as a promising alternative. There is one critical point though that seems a hard requirement that this won't cover. We would like to have the `blob/close` + `assert/location claim`  to happen right away. While they do not happen the data is not available. This is a requirement that we have continuously not able to accomplish, and we always had reports of delays to get data ready to read (today in w3up is due to delay of replicator to CF).

The problem is:

> When service is made aware of this it can then perform blob/close or something along those lines which is more or less what store/confirm is. Here is the rough workflow diagram

By being the client to immediately tell that seems a more efficient way than service having a huge list of tasks to handle in a queue. Also ends up being less prone to errors and maintenance on failures.

We could potentially consider some event subscription and associated runs with the receipt reporting API. Even though this will also require us to thing on how we want to manage receipts being sent by client.

Reply:
> By being the client to immediately tell that seems a more efficient way than service having a huge list of tasks to handle in a queue. Also ends up being less prone to errors and maintenance on failures, and obviously way more expensive to run.

I think we need to do both (because we have no guarantee client will ever notify us).

1. Have run scheduled on pre-signed url expiry in case we have not got any update from the clients.
2. Have a handler on the receipt that will perform the closure.

> We could potentially consider some event subscription and associated runs with the receipt reporting API. Even though this will also require us to thing on how we want to manage receipts being sent by client.

ucanto agent messages are send both ways and they [can hold both invocations and receipts](
https://github.com/web3-storage/ucanto/blob/c24a99c3078b4c82f3133c5c1bcda6aa63667c84/packages/interface/src/lib.ts#L800-L821) which is to say that client can send receipt to a server after they have completed an upload.

You are correct right now our service router will not react to received receipts in any way, however we could [hook server up to do things with `incoming.report` which is currently ignored](https://github.com/web3-storage/w3infra/blob/main/upload-api/functions/ucan-invocation-router.js#L276-L277)

I think trickiest bit here would be to track dependencies that is `blob/close` would have to execute when we get receipt for `blob/write`. However we can even pun on that and simply add metadata in the receipt that will signal us to poll `blob/write`. Which is not perfect by any means but IMO better than to have no trace of the workflow capture in the receipts.

Alternatively we could just have a endpoint or even a capability to poll a task where handler will attempt to check dependencies and if they are available then run the task.

Both approaches would be general to all of the cases where we have workflows with non-linear dependency graphs as oppose to being on off workarounds for not having support for this yet.

> I would like to get closer to that, but would like to figure out how to not compromise better service for users and service cost

That is fare! I think there is a possible compromise here where we can do a just a bit more, not much and get workflows with traceable receipts for it. If we really want to aggressively cut scope we could even device something like

```js
{
   can: "system/poll",
   with: "did:key:zAlice",
   nb: {
      task: { '/': 'bafy...pending' },
      receipts: {
         'bafy...task': { '/'; 'bafy..receipt' }
      }
   }
}
```

Where handler will effectively just try to run the task we already created in the and put in the receipt and provide all the dependencies for it.

Reply:
while we decided to go to simplest solution now, I think there is a lot of things here that indeed could be the future. We should save this info in an issue, maybe within the context of the `blob/*` proposal you put together?

For instance, the Job on expiration seems also something we should do, but I would like to not increase the scope here of dataanywhere, which by itself is already quite big

Reply:
@Gozala I think we can frame this differently to better be aligned toward implementing what you suggest later on, once we can. If we frame this into the context of the proof of delivery, where client will send a challenge that service needs to fulfil, this interaction as is needs to happen anyway. We can leverage it for the time being for submitting the index, until we have ease to implement this once UCAN updates land and we bubble them up